rcl/include/rcl/client.h

```diff
@@ -0,0 +1,300 @@
+ *
+ *    #include <rosidl_generator_cpp/service_type_support.hpp>
+ *    #include <example_interfaces/srv/add_two_ints.h>
+ *    rosidl_service_type_support_t * ts = rosidl_generator_cpp::get_service_type_support_handle<
+ *      example_interfaces::srv::AddTwoInts>();
+ *
+ * The rosidl_service_type_support_t object contains service type specific
+ * information used to send or take requests and responses.
+ *
+ * \TODO(wjwwood) update this once we've come up with an official scheme.
+ * The topic name must be a non-empty string which follows the topic naming
+ * format.
+ *
+ * The options struct allows the user to set the quality of service settings as
+ * well as a custom allocator which is used when initializing/finalizing the
+ * client to allocate space for incidentals, e.g. the service name string.
+ *
+ * Expected usage (for C services):
+ *
+ *    #include <rcl/rcl.h>
+ *    #include <rosidl_generator_c/service_type_support.h>
+ *    #include <example_interfaces/srv/add_two_ints.h>
+ *
+ *    rcl_node_t node = rcl_get_zero_initialized_node();
+ *    rcl_node_options_t node_ops = rcl_node_get_default_options();
+ *    rcl_ret_t ret = rcl_node_init(&node, "node_name", &node_ops);
+ *    // ... error handling
+ *    rosidl_service_type_support_t * ts = ROSIDL_GET_MESSAGE_TYPE_SUPPORT(
+ *      example_interfaces, AddTwoInts);
+ *    rcl_client_t client = rcl_get_zero_initialized_client();
+ *    rcl_client_options_t client_ops = rcl_client_get_default_options();
+ *    ret = rcl_client_init(&client, &node, ts, "chatter", &client_ops);
+ *    // ... error handling, and on shutdown do finalization:
+ *    ret = rcl_client_fini(&client, &node);
+ *    // ... error handling for rcl_client_fini()
+ *    ret = rcl_node_fini(&node);
+ *    // ... error handling for rcl_deinitialize_node()
+ *
+ * This function is not thread-safe.
+ *
+ * \param[inout] client preallocated client structure
+ * \param[in] node valid rcl node handle
+ * \param[in] type_support type support object for the topic's type
+ * \param[in] service_name the name of the service to request
+ * \param[in] options client options, including quality of service settings
+ * \return RCL_RET_OK if the client was initialized successfully, or
+ *         RCL_RET_NODE_INVALID if the node is invalid, or
+ *         RCL_RET_ALREADY_INIT if the client is already initialized, or
+ *         RCL_RET_INVALID_ARGUMENT if any arguments are invalid, or
+ *         RCL_RET_BAD_ALLOC if allocating memory fails, or
+ *         RCL_RET_ERROR if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_client_init(
+  rcl_client_t * client,
+  const rcl_node_t * node,
+  const rosidl_service_type_support_t * type_support,
+  const char * service_name,
+  const rcl_client_options_t * options);
+
+/// Finalize a rcl_client_t.
+/* After calling, the node will no longer be advertising that it is publishing
+ * on this topic (assuming this is the only client on this topic).
```

Comment:
This sentence doesn't look right. I'm not 100% sure what it should say, I guess that the node is no longer calling a particular service.